src/Tools/dotnet-monitor/Egress/S3/MultiPartUploadStream.cs

```diff
@@ -0,0 +1,147 @@
+        if (_offset == 0 || _offset < MinimumSize)
+            return;
+        await DoWriteAsync(cancellationToken);
+    }
+
+    public async Task FinalizeAsync(CancellationToken cancellationToken)
+    {
+        if (Disposed)
+            throw new ObjectDisposedException(nameof(MultiPartUploadStream));
+        if (_offset == 0)
+            return;
+        await DoWriteAsync(cancellationToken);
+    }
+
+    public override void Flush()
+    {
+        throw new NotSupportedException();
+    }
+
+    public override int Read(byte[] buffer, int offset, int count)
+    {
+        throw new NotSupportedException();
+    }
+
+    public override long Seek(long offset, SeekOrigin origin)
+    {
+        throw new NotSupportedException();
+    }
+
+    public override void SetLength(long value)
+    {
+        throw new NotSupportedException();
+    }
+
+    public override async Task WriteAsync(byte[] buffer, int offset, int count, CancellationToken cancellationToken)
+    {
+        if (Disposed)
+            throw new ObjectDisposedException(nameof(MultiPartUploadStream));
+
+        int BytesAvailableInBuffer() { return _buffer.Length - _offset; }
+        do
+        {
+            int bytesToCopy = Math.Min(count, BytesAvailableInBuffer());
+            Array.Copy(buffer, offset, _buffer, _offset, bytesToCopy);
+            offset += bytesToCopy; // move offset of part buffer
+            count -= bytesToCopy; // reduce amount of bytes which still needs to be written
+            _position += bytesToCopy; // move global position
+
+            // part buffer is full -> trigger upload of part
+            if (BytesAvailableInBuffer() == 0)
+                await DoWriteAsync(cancellationToken);
+        } while (count > 0);
+    }
+
+    private async Task DoWriteAsync(CancellationToken cancellationToken)
+    {
+        await using var stream = new MemoryStream(_buffer, 0, _offset);
+        stream.Position = 0;
+        var uploadRequest = new UploadPartRequest
+        {
+            BucketName = _bucketName,
+            Key = _objectKey,
+            InputStream = stream,
+            PartSize = _offset,
+            UploadId = _uploadId,
+            PartNumber = _parts.Count
+        };
+        var response = await _client.UploadPartAsync(uploadRequest, cancellationToken);
+        _parts.Add(new PartETag(response.PartNumber, response.ETag));
+        _offset = 0;
+    }
+
+    public override void Write(byte[] buffer, int offset, int count)
+    {
+        throw new NotSupportedException();
+    }
+
+    public override bool CanRead => false;
+    public override bool CanSeek => false;
+    public override bool CanWrite => !Disposed;
+    public override long Length => throw new NotSupportedException();
+
+    public override long Position
+    {
+        get => _position;
+        set => throw new NotSupportedException();
+    }
+
+    protected override void Dispose(bool disposing)
```

Comment:
Consider changing this to override the `Close` method (which is called by `Dispose` and indirectly by `DisposeAsync`) so others who look at this don't wonder why `Dispose` was implemented but not `DisposeAsync`.